common/src/pack.rs

```diff
@@ -0,0 +1,176 @@
+//! pack.rs defines utilities for serializing Solana accounts to/from bytes.
+
+use solana_sdk::program_error::ProgramError;
+
+/// The Pack trait defines Account serialization for Solana programs.
+///
+/// If possible, don't use `*_unchecked` methods.
+pub trait Pack<'a>: serde::Serialize + serde::Deserialize<'a> {
+    /// Serializes `src` into `dst`. The size of the serialization and
+    /// dst must be equal.
+    fn pack(src: Self, dst: &mut [u8]) -> Result<(), ProgramError>;
+
+    /// Deserializes `src` into Self. The deserialized object need not
+    /// use all the bytes in `src`. This is the case, for example,
+    /// when encoding variable length data, say, when one has a src
+    /// array of all zeroes, and a Self that has a Vec<u64>.
+    ///
+    /// Use care when using this directly. If using fixed length structs
+    /// always use the `unpack` method, instead.
+    fn unpack_unchecked(src: &[u8]) -> Result<Self, ProgramError>;
```

Comment:
I think `src` should be an `&mut &[u8]`, and should be narrowed to contain only the unconsumed bytes when the function returns.